test/utils/update_resources.go

```diff
@@ -33,10 +35,34 @@ const (
 	waitRetryTImeout    = 5 * time.Minute
 )
 
-func ScaleResourceWithRetries(scaler kubectl.Scaler, namespace, name string, size uint, gr schema.GroupResource) error {
-	waitForScale := kubectl.NewRetryParams(updateRetryInterval, updateRetryTimeout)
-	waitForReplicas := kubectl.NewRetryParams(waitRetryInterval, waitRetryTImeout)
-	if err := scaler.Scale(namespace, name, size, nil, waitForScale, waitForReplicas, gr); err != nil {
```

Comment:
(sorry I thought I submitted this comment from my phone.. but seems like I didn't)

Reply:
I already wrote that comment somewhere before.
There are multiple reasons:
1. we want it to fail after waitRetryTimeout (not potentially after 2x longer time, which could happen in that time)
2. if scaling the object succeeded and we just failed waiting for pods, we should start from the beginning and again scale the object - we already did that.

I asked Krzysiek explicitly to do it this way.

Reply:
I don't think that's true.

> if scaling the object succeeded and we just failed waiting for pods, we should start from the beginning and again scale the object - we already did that.

I've also considered this, and FMU it doesn't matter. Even if we fail in the 'waiting' part and retry the whole operation that's fine, because then the 'scale' would just be a no-op if we try to set it to the same replicas again. See:

```
$ kubectl get rc -n e2e-tests-load-30-nodepods-44-l9njz load-small-9993 -o yaml
...
resourceVersion: "12120458"
replicas: 2

$ kubectl scale -n e2e-tests-load-30-nodepods-44-l9njz --replicas=2 rc/load-small-9993
replicationcontroller/load-small-9993 scaled

$ kubectl get rc -n e2e-tests-load-30-nodepods-44-l9njz load-small-9993 -o yaml
...
resourceVersion: "12120458"
replicas: 2

$ kubectl scale -n e2e-tests-load-30-nodepods-44-l9njz --replicas=2 rc/load-small-9993 --v=8
GET http://localhost:8080/api/v1/namespaces/e2e-tests-load-30-nodepods-44-l9njz/replicationcontrollers/load-small-9993
^ Only GET, no subsequent write call happens
```

> we want it to fail after waitRetryTimeout (not potentially after 2x longer time, which could happen in that time)

We don't wait for 2x longer. We should fail if there was 'any timeout' (either during scaling or waiting), which is correct behaviour IMO. However, if there's some retriable error, we'll redo the whole operation.

In short, I think this PR should be a 2-line change:

```
if err != nil && err != wait.Timeout && IsRetryable(err); then retry; else fail
```

Reply:
> We don't wait for 2x longer. We should fail if there was 'any timeout' (either during scaling or waiting), which is correct behaviour IMO. However, if there's some retriable error, we'll redo the whole operation.

That isn't true. If 'waiting for pods will fail right before timeout on some other error, then:
- that function will return with an error
- the external "wait.Poll" will still be before timeout, so it will trigger the `Scale` one more time
- the new Sclae will start with the original timeouts from scratch

Yes - we can start adjusting timeouts ourselves, but I really don't like this option.

> because then the 'scale' would just be a no-op if we try to set it to the same replicas again. See:

Yes I know it no-op and it won't even hit etcd because we check whether what we are trying to write is exactly the same or not.
But still - that is unnecessary call to apiserver, which I don't want to send.

But the first argument is much more important here.

Reply:
I think given that this is just test code for retrying we probably don't need to be so critical with the retry logic. Yes, if we fail in the waiting step, we'll potentially add:
- 1 extra GET call during next scale step (maybe some more if the get fails)
- 1 extra round of waiting for max 5 mins

But given that we shouldn't be needing this retry logic for >99% of the calls likely, the above shouldn't cause much difference imo. Btw - I'm also willing to change waitRetryTimeout to half (i.e 160s) as this is the only purpose for it's use and that should already offer enough retries (= 5) :)

One more thing - the cost of duplicating the whole function (and keeping it in sync with kubectl) is probably more than the benefit I feel.

Reply:
> 1 extra round of waiting for max 5 mins

That changes our expectations - something that was failing previously may pass now - we may hide potential regressions.

> One more thing - the cost of duplicating the whole function (and keeping it in sync with kubectl) is probably more than the benefit I feel.

In this particular case I don't agree - we the current refactoring we pretty much don't have any significnat duplication - there are two building block function that we are reusing  - the glueing code is trivial in this case.

I have a very strong opinion that this approach is better.

Reply:
> That changes our expectations - something that was failing previously may pass now - we may hide potential regressions.

That's why I was suggesting halving the timeout (to keep wait the same). With that we should catch at least those regressions we're catching now.

> there are two building block function that we are reusing - the glueing code is trivial in this case

Yes, that was reason why I had the comment on moving wait to separate function. IMO there's some machinery there we're unnecessarily duplicating (e.g initializing `preconditions`, using some function signatures, etc).

I wouldn't insist, but when I think of it, it seems like we're making a 60-line change instead of a 3-line change just because we don't want to half the timeout (which is only used by this function :)

Reply:
I don't want to half the timeout just to make this PR smaller.
If we believe that half of the timeout is enough, let's change that independently and we that we're back to the same problem. This PR is not supposed to change the default behavior (in case on errors).

Reply:
To begin with, I'm not sure what's so specific with that number (iiuc it's sth that we just informally agreed as a reasonable number).

> This PR is not supposed to change the default behavior (in case on errors).

Who decides what's the default behaviour (besides us)? :)
Just because we've been having that for a while, doesn't mean we can't change it. For e.g if we can catch more regressions by changing it, then we probably should.

Also IIUC from a user perspective they'll call `kubectl scale --replicas=X --wait-for-replicas` and if it fails they'll just call it again. Wdyt?

Reply:
> To begin with, I'm not sure what's so specific with that number (iiuc it's sth that we just informally agreed as a reasonable number).

Nothing specific. But if we agree that X is a reasonable number, the we should use X, not "something random between X and 2X".

> For e.g if we can catch more regressions by changing it, then we probably should.

I agree. But by doing what you suggest we can only catch less regressions, not more.

> Also IIUC from a user perspective they'll call kubectl scale --replicas=X --wait-for-replicas and if it fails they'll just call it again.

If you're a user, then you know how long you want to wait. If you're programming (which we do), we want to have a well defined timeout.

Reply:
> But if we agree that X is a reasonable number, the we should use X, not "something random between X and 2X".

That's not true. Currently we're randomly waiting for "sth b/w 0 and X", with my suggestion it'll be "sth b/w 0 and 2X" :)

> But by doing what you suggest we can only catch less regressions, not more.

Why do you say so? We still continue to retry for max 5 minutes (but split across 2 cycles).

> If you're a user, then you know how long you want to wait. If you're programming (which we do), we want to have a well defined timeout.

My point was not about timeout, but the operation granularity at which we retry. Btw - I think some users would want to exercise timeout (using --timeout flag kubectl exposes), for e.g while doing a large scale-up operation?

Reply:
> That's not true. Currently we're randomly waiting for "sth b/w 0 and X", with my suggestion it'll be "sth b/w 0 and 2X" :)

Well - depends how you're looking into that. Yes - if some error happens, we may end up before X. But there is hard timeout of X.
With your suggestion there won't be hard timeout. Depending what happened in the meantime, the operation may fail *with timeout error* after anything between [X, 2X] - that's pretty bad experience.

> Why do you say so? We still continue to retry for max 5 minutes (but split across 2 cycles).

That isn't true. Imagine a bug that apiserver is crashing every 5 minutes. And imagine that now we regressed and 1% of RCs take more than 5m to scale.
With my change, we will discover it deterministacally. With your, we may or may not discover it.

> My point was not about timeout, but the operation granularity at which we retry. Btw - I think some users would want to exercise timeout (using --timeout flag kubectl exposes), for e.g while doing a large scale-up operation?

Yes, but if it fails after almost timeout (with some random error) - what they do? If they were waiting for 10 minutes, they probably won't wait for another 10. And I don't want to play with timeouts.

Reply:
> With your suggestion there won't be hard timeout

There is a hard timeout, it's 2X. To clarify - with my change, we're just splitting the wait across two iterations (as X + X) instead of waiting in single iteration for 2X.
And halving "waitRetryTimeout" will achieve the same effect.

> With your, we may or may not discover it.

We will detect it with my change. We just need to adjust overall timeout to 5 mins :)

Btw - I'm fine with the current change. I don't mind much that the size of PR is big. But if we can achieve similar effect with a 3-line change, just saying that I'd prefer it more.

Reply:
> There is a hard timeout, it's 2X. To clarify - with my change, we're just splitting the wait across two iterations (as X + X) instead of waiting in single iteration for 2X.
And halving "waitRetryTimeout" will achieve the same effect.

No - there is significantly different UX with your proposal which in my opinion is wrong.
With the current PR:
- if you get "timeout" error" then we know it took X
With your proposal,
- if you get "timeout" error, then it took some time between X and 2X
That is really misleading.

> We will detect it with my change. We just need to adjust overall timeout to 5 mins :)

But that means we need to start playing with timeouts. So when the internal function (Scale) fails with error, we need to call it with a timeout that is original timeout minus time that already passed.
That's something I want to avoid - I prefer some small duplication of code that is fairly trivial.

> But if we can achieve similar effect with a 3-line change, just saying that I'd prefer it more.

What I claim is that we can't achieve the same.

Reply:
That sounds reasonable. Yes, there's some UX difference, but given that it's just retry logic in our test I felt we probably don't need to worry too much about that.

Anyway, let's get this in to stop our flakes.